sky/data/sky_csync.py

```diff
@@ -0,0 +1,545 @@
+    """Runs daemon to sync the source to the bucket every INTERVAL seconds.
+
+    Creates an entry of pid of the sync process in local database while sync
+    command is running and removes it when completed.
+
+    Args:
+        source (str): The local path to the directory that you want to sync.
+        storetype (str): The type of cloud storage to sync to.
+        destination (str): The bucket or subdirectory in the bucket where the
+            files should be synced.
+        num_threads (int): The number of threads to use for the sync operation.
+        interval_seconds (int): The time interval, in seconds, at which to run
+            the sync operation.
+        delete (bool): Whether or not to delete files in the destination that
+            are not present in the source.
+        no_follow_symlinks (bool): Whether or not to follow symbolic links in
+            the source directory.
+    """
+    mountpoint_path = os.path.abspath(os.path.expanduser(source))
+    # if the given source is already mounted with CSYNC, terminate it.
+    if _get_csync_pid_from_mountpoint_path(mountpoint_path):
+        _terminate([mountpoint_path])
+    csync_pid = os.getpid()
+
+    # create directories which mountpoint_path redirects
+    # writing and reading to.
+    csync_write_path = os.path.abspath(
+        os.path.expanduser(_CSYNC_WRITE_PATH.format(pid=csync_pid)))
+    csync_read_path = os.path.abspath(
+        os.path.expanduser(_CSYNC_READ_PATH.format(pid=csync_pid)))
+    if os.path.isdir(csync_write_path):
+        shutil.rmtree(csync_write_path)
+    if os.path.isdir(csync_read_path):
+        shutil.rmtree(csync_read_path)
+    os.makedirs(csync_write_path, exist_ok=True)
+    os.makedirs(csync_read_path, exist_ok=True)
+
+    # mounting cloud object storage on the read path
+    with tempfile.NamedTemporaryFile(mode='w+', delete=False) as script_file:
+        storage_mount_script = get_storage_mount_script(storetype, destination,
+                                                        csync_read_path)
+        script_file.write(storage_mount_script)
+        # ensure all data is written to the file
+        script_file.flush()
+        storage_fuse_mount_cmd = f'bash {script_file.name}'
+        storage_fuse_mount_pid, rc, _ = _handle_fuse_process(
+            storage_fuse_mount_cmd)
+    if rc != 0:
+        sys.exit(exceptions.CSYNC_TERMINATE_FAILURE_CODE)
+
+    # redirect read/write of mountpoint_path by mounting redirection FUSE
+    redirect_mount_cmd = mounting_utils.get_redirect_mount_cmd(
+        mountpoint_path, csync_read_path, csync_write_path)
+    redirect_fuse_mount_pid, rc, _ = _handle_fuse_process(redirect_mount_cmd)
+    if rc != 0:
+        sys.exit(exceptions.CSYNC_TERMINATE_FAILURE_CODE)
+
+    _add_running_csync(csync_pid, storage_fuse_mount_pid,
+                       redirect_fuse_mount_pid, mountpoint_path)
+    sys.stdout.write('Successfully ran FUSE mounting processes.')
```

Comment:
is there any special reason to use `sys.stdout.write` all over the file? why not use logging?